nativelib/src/main/scala/java/lang/ieee754tostring/ryu/RyuDouble.scala

```diff
@@ -0,0 +1,548 @@
+// Copyright 2018 Ulf Adams
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Credits:
+//
+//    1) This work is a heavily modified derivation of the original work by
+//       Ulf Adams at URL: https://github.com/ulfjack/ryu.
+//       As such, it inherits the Apache license of the original work.
+//       Thank you Ulf Adams.
+//
+//    2) The original java sources were converted to a rough draft of
+//       scala code using the service at URL: javatoscala.com.
+//
+//       The raw conversion did not compile and contained bugs due
+//       to the handling of break and return statements, but it saved
+//       days, if not weeks, of effort.
+//
+//       Thank you javatoscala.com.
+//
+//    3) All additional work, including introduced bugs,  is an original
+//       contribution to Scala Native development.
+
+package java.lang.ieee754tostring.ryu
+
+import java.math.BigInteger
+
+import RyuRoundingMode._
+
+object RyuDouble {
+
+  var DEBUG: Boolean = false
+
+  private val DOUBLE_MANTISSA_BITS: Int = 52
+
+  private val DOUBLE_MANTISSA_MASK: Long = (1L << DOUBLE_MANTISSA_BITS) - 1
+
+  private val DOUBLE_EXPONENT_BITS: Int = 11
+
+  private val DOUBLE_EXPONENT_MASK: Int = (1 << DOUBLE_EXPONENT_BITS) - 1
+
+  private val DOUBLE_EXPONENT_BIAS: Int = (1 << (DOUBLE_EXPONENT_BITS - 1)) - 1
+
+  private val POS_TABLE_SIZE: Int = 326
+
+  private val NEG_TABLE_SIZE: Int = 291
+
+  // Only for debugging.
+  private val POW5: Array[BigInteger] = new Array[BigInteger](POS_TABLE_SIZE)
+
+  private val POW5_INV: Array[BigInteger] =
+    new Array[BigInteger](NEG_TABLE_SIZE)
+
+  private val POW5_BITCOUNT: Int = 121 // max 3*31 = 124
+
+  private val POW5_QUARTER_BITCOUNT: Int = 31
+
+  private val POW5_SPLIT: Array[Array[Int]] =
+    Array.ofDim[Int](POS_TABLE_SIZE, 4)
+
+  private val POW5_INV_BITCOUNT: Int = 122 // max 3*31 = 124
+
+  private val POW5_INV_QUARTER_BITCOUNT: Int = 31
```

Comment:
Please use `final val CONSTANT = v` for numeric constants (no type ascription).

Reply:
Private modifier can be omitted because the whole thing is part of runtime which is "private" package by definition.

Reply:
Done, thank you for the suggestion.

"private" removed from all variable and method declarations:

Reply:
If you don't have a Long ascription, scalac will inline them at use site, which is a good thing.

Reply:
Done, as requested.

I left private on all method declarations other than the one toString method I want to be
public. Does that help the SN compiler generate less code (I hope) or is it a mistake?

Reply:
`private` modifier doesn't really matter from the performance point of view, it's purely a question of API guarantees from an external point of view. The reason I suggested dropping private is that runtime module is "internals" of the SN, so use of any of those APIs is "at your own risk" and we don't guarantee any compatibility between releases.

Reply:
Thank you for your patient explanation.